src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,494 @@
+ * {@code addSalt} methods may be called repeatedly (and chained). This provides
+ * for use-cases where a portion of the IKM resides in a non-extractable
+ * {@code SecretKey} and the whole IKM cannot be provided as a single object.
+ * The caller may wish to provide a label (or other components) of
+ * the IKM without having access to all portions. The same feature is
+ * available for salts.
+ * <p>
+ * The above feature is particularly useful for "labeled" HKDF Extract used in
+ * TLS 1.3 and HPKE, where the IKM consists of concatenated components, which
+ * may include both byte arrays and (possibly non-extractable) secret keys.
+ * <p>
+ * Examples:
+ * {@snippet lang = java:
+ *
+ * // this usage depicts the initialization of an HKDF-Extract AlgorithmParameterSpec
+ * AlgorithmParameterSpec derivationSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(label)
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).extractOnly();
+ *
+ *
+ *}
+ * {@snippet lang = java:
+ *
+ * // this usage depicts the initialization of an HKDF-Expand AlgorithmParameterSpec
+ * AlgorithmParameterSpec derivationSpec =
+ *             HKDFParameterSpec.expandOnly(prk, info, 32);
+ *
+ *}
+ * {@snippet lang = java:
+ *
+ * // this usage depicts the initialization of an HKDF-ExtractExpand AlgorithmParameterSpec
+ * AlgorithmParameterSpec derivationSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).thenExpand(info, 32);
+ *
+ *}
+ *
+ * @see javax.crypto.KDF
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public interface HKDFParameterSpec extends AlgorithmParameterSpec {
+
+    /**
+     * This {@code Builder} builds {@code Extract} and {@code ExtractThenExpand}
+     * objects.
+     * <p>
+     * The {@code Builder} is initialized via the {@code ofExtract} method of
+     * {@code HKDFParameterSpec}. As stated in the class description,
+     * {@code addIKM} and/or {@code addSalt} may be called as needed. Finally,
+     * the object is "built" by calling either {@code extractOnly} or
+     * {@code thenExpand} for {@code Extract} and {@code ExtractThenExpand}
+     * use-cases respectively. Note that the {@code Builder} is not
+     * thread-safe.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class Builder {
+
+        List<SecretKey> ikms = new ArrayList<>();
+        List<SecretKey> salts = new ArrayList<>();
+
+        private Builder() {}
+
+        /**
+         * Creates a {@code Builder} for an {@code Extract}.
+         *
+         * @return a {@code Builder} to mutate
+         */
+        private Builder createBuilder() {
+            return this;
+        }
+
+        /**
+         * Builds an {@code Extract} from the current state of the
```

Comment:
Is it better to say "Builds an Extract object"? Same with the `@return` line and other methods.